pkg/reconciler/ingress/ingress.go

```diff
@@ -467,6 +467,10 @@ func getLBStatus(gatewayServiceURL string) []v1alpha1.LoadBalancerIngressStatus
 	}
 }
 
-func enableReconcileGateway(ctx context.Context) bool {
-	return config.FromContext(ctx).Network.AutoTLS || config.FromContext(ctx).Istio.ReconcileExternalGateway
+func (r *Reconciler) shouldReconcileTLS(ia *v1alpha1.Ingress) bool {
+	// For the case which the finalizer was added to an Ingress, we should recocnile TLS.
+	// This is because IngressTLS could be added before and then deleted later. We
+	// need to make sure deleting IngressTLS also cleans up the TLS of the Gateway.
+	return (len(ia.GetFinalizers()) != 0 && ia.GetFinalizers()[0] != r.finalizer) ||
+		(ia.IsPublic() && len(ia.Spec.TLS) > 0)
```

Comment:
I don't get the public part.  If I issue a cluster CA signed cert for a `.svc.cluster.local` address, shouldn't we terminate it with the provided secret?

Reply:
Currently IngressTLS is only reflected on public Gateway https://github.com/knative/serving/blob/master/pkg/reconciler/ingress/ingress.go#L197.
Ideally we may want to choose the target Gateway (public V.S. cluster local) based on Ingress's visibility.

I would like to keep `IsPublic` temporarily in this PR, and think about removing it later with more consideration.

Reply:
> > I am working on a test case which an empty IngressTLS will clean up the TLS of Gateway.
> 
> As part of this PR?

Yes. Just added the unit test case into this PR